# radseqQC
Pipeline for QC of RAD-seq data

[![Build Status](https://travis-ci.org/remiolsen/radseqQC.svg?branch=master)](https://travis-ci.org/remiolsen/radseqQC)
[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A50.30.0-brightgreen.svg)](https://www.nextflow.io/)

[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg)](http://bioconda.github.io/)
[![Docker](https://img.shields.io/docker/automated/remiolsen/radseqQC.svg)](https://hub.docker.com/r/remiolsen/radseqQC)
![Singularity Container available](
https://img.shields.io/badge/singularity-available-7E4C74.svg)


### Introduction
radseqQC: Pipeline for QC of RAD-seq data

A work-in-progress pipeline for QC of RAD-seq data generated by National Genomics Infrastructure in Stockholm. The pipeline was was primarily made for an in-house RAD-seq protocol that resembles ezRAD, however other flavours of RAD-seq might work as long as [Stacks](http://catchenlab.life.illinois.edu/stacks/) support them. RAD-seq allows for deep yet sparsely sampled sequencing of many individuals in a highly multiplexed manner, where typical applications includes QTL mapping, GWAS studies, high resolution population differentiation/phylogeny, pedigree reconstruction and SNP discovery for other more high throughput assays. These steps are mainly designed to only characterize the data and attempt to correct defects (e.g. adapter contamination and restriction-site sequencing errors) for further downstream analysis, but not to draw any biologically relevant conclusions:

* FastQC
* Adapter trimming using trimmomatic
* Stacks de novo pipeline with the most lenient parameters


The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It comes with docker / singularity containers making installation trivial and results highly reproducible.


### Documentation
The radseqQC pipeline comes with documentation about the pipeline, found in the `docs/` directory:

1. [Installation](docs/installation.md)
2. Pipeline configuration
    * [Local installation](docs/configuration/local.md)
    * [Adding your own system](docs/configuration/adding_your_own.md)
3. [Running the pipeline](docs/usage.md)
4. [Output and how to interpret the results](docs/output.md)
5. [Troubleshooting](docs/troubleshooting.md)

### Credits
This pipeline was written by Remi-Andre Olsen ([remiolsen](https://github.com/remiolsen)) at [SciLifeLab](http://www.scilifelab.se).
